### ISSUE TITLE
How does quanto calibrate torch functions?

### ISSUE THREAD

**Issue by shuokay** [huggingface/quanto#152]
I have learned quanto calibrate ops in module forms by adding module hooks, but how about torch functions like `torch.sigmoid`, `torch.elu`, and `torch.log` etc?
I think the output scale of `torch.sigmoid` could be directly evaluated similarly to quanto's approach with `softmax`. Additionally, `torch.elu` might be substituted with `torch.nn.ELU`.
However, I'm uncertain how functions like `torch.log`, which are unbounded and lack explicit module forms will be calibrated within quanto.

**Comment by dacorvo** [comment]
It doesn't. Only the inputs/outputs of modules can be calibrated.
I had a branch where I managed to calibrate around functions using a DispatchMode but it was a bit overkill since most of the time the next function does not accept quantized inputs, so instead I used it to deactivate the quantization of module outputs when they a re not consumed by a compatible operation (streamline mode).

**Comment by shuokay** [comment]
@dacorvo Can you share the branch? Is it tracking_mode branch here: https://github.com/huggingface/quanto/tree/tracking_mode?

**Comment by dacorvo** [comment]
I can't find it: it was an earlier version of this, going in a different direction

**Comment by dacorvo** [comment]
And just to be clear, this is not something I plan to change in the near future: advanced dispatch is only required for activations, and for now the quantization of activations is not really required, since:
- only operations using int8 activations can be accelerated,
- in most transformers models int8 outputs are almost immediately dequantized because they are used in incompatible operations.
IMHO, only float8 activations are interesting, as they will be compatible with much more operations.
But this will be usable only once we can use accelerated float8 operations.

**Comment by dacorvo** [comment]
quanto is a pytorch quantization toolkit, producing pytorch quantized models to be deployed on hosts running pytorch.
More, it is designed to immediately dequantize tensors whenever it reaches an operation that is incompatible with the underlying storage data type. For integer tensors, this means very few operations indeed (see under tensor/ops).

If you want to target NPUs, you can use quanto to obtain quantized weights, but you will never be able to replace in the pytorch graph the operations that are not compatible with integer arithmetics. There are other tools to do that, one of them being [quantizeml](https://doc.brainchipinc.com/api_reference/quantizeml_apis.html), one of my previous projects (closed-source unfortunately).
